Update the RTC-S1 SAS for the cal/val delivery

gshiroma opened the pull request:
This PR updates the RTC-S1 SAS for the cal/val delivery:
- Add coverage check for ancillary inputs (DEM);
- Provide three mosaicking options: `average`, `first`, and `bursts_center`;
- Reproject burst product before mosaicking if needed;
- Remove `/science/SENTINEL1` base path from RTC-S1 HDF5 metadata;
- Rename HDF5 imagery group `grids` to `data`.

Minor updates:
- Substitute metadata field `ISCEVersion` to `isce3Version`;
- Update some comments, fix typos, indentation, and bugs;
- Update `dem_description` to `dem_file_description`;
- Update default compression algorithm from `ZSTD` to `DEFLATE`;
- Separate runconfig burst and mosaic geogrids;
- Update functions to compute burst and mosaic geogrids;
- Update metadata field `orbitType`;
- Remove requirement of `save_nlooks = True`;
- Add test runconfig at `tests/runconfigs/s1b_los_angeles.yaml` and use it for workflow testing rather than using the runconfig  downloaded from zenodo (during workflow test).

seongsujeong commented:
@gshiroma This is a huge PR as you mentioned. Please give me the highlights about what feature(s) are added and changed. That will save a lot of time on my end.
You mentioned "Add coverage check for ancillary inputs (DEM)". One of things I remember is the mosaicking function. Anything else?

gshiroma commented:
> @gshiroma This is a huge PR as you mentioned. Please give me the highlights about what feature(s) are added and changed. That will save a lot of time on my end. You mentioned "Add coverage check for ancillary inputs (DEM)". One of things I remember is the mosaicking function. Anything else?

Just updated the PR description. Thank you, @seongsujeong

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+       tile_min_x_utm: float
+              Tile minimum X-coordinate (UTM)
+       tile_max_x_utm: float
+              Tile maximum X-coordinate (UTM)
```
Based on burst database, some high-latitude area uses polar stereographic. Therefore I think we should use more general term than `UTM`. How about replacing `UTM` into `map coordinate?`

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+
+
+
+def check_ancillary_inputs(check_ancillary_inputs_coverage,
```
I guess this is basically copied from PROTEUS. I would suggest to modify the terminology accordingly to RTC.
In specific, if I understand this correctly, `tile` should be the geogrid of output.

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+def check_ancillary_inputs(check_ancillary_inputs_coverage,
+                           dem_file, geogrid,
+                           metadata_dict, logger=None):
+    """Check for the existence and coverage of provided ancillary inputs
+       (DEM) wrt. to the HLS tile. The function
+       also updates the product's dictionary metadata indicating the coverage
+       of each ancillary input wrt. to the input HLS tile
+
```
The description here needs to be updated for in accordance with RTC.

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+        # Create input ancillary polygon
+        file_polygon = _get_ogr_polygon(min_x, max_y, max_x, min_y, file_srs)
+
+        coverage_logger_str = file_description+' coverage'
+        coverage_metadata_str = file_type+'_COVERAGE'
```
```suggestion
        coverage_logger_str = file_description + ' coverage'
        coverage_metadata_str = file_type + '_COVERAGE'
```

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+        # outside of the tile grid.
+        gdal_ds = gdal.Open(file_name, gdal.GA_ReadOnly)
+
+        file_geotransform = gdal_ds.GetGeoTransform()
```
What do you think about sth. like `anc_file_geotransform` to make it clear that this is the information from ancillary input? (and other variable names that starts with `file` as well)

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+        if _antimeridian_crossing_requires_special_handling(
+                file_srs, min_x, tile_min_x, tile_max_x):
+
+            logger.info(f'The input HLS product crosses the antimeridian'
```
```suggestion
            logger.info(f'The input RTC product crosses the antimeridian'
```

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+
+        del gdal_ds
+
+        max_x = min_x + file_width * dx
+        min_y = max_y + file_length * dy
```
it would be great if the variable name more clearly mention this is from ancillary input data.

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+                        f' {file_description}: {file_name}')
+
+            # Left side of the antimeridian crossing: -180 -> +180
+            file_polygon_1 = _get_ogr_polygon(-180, 90, max_x, -90, file_srs)
```
I think this line assumes that the ancillary file is in geographic coordinates. Can we always assume that the user is going to provide the DEM whose projection is sth. like EPSG 4326? What is going to happen when the input DEM is in UTM?

gshiroma replied:
Hi @seongsujeong , we first test if the ancillary file is in geographic coordinates within the function `_antimeridian_crossing_requires_special_handling()`, specifically in the test flag `flag_input_geographic_and_longitude_not_0_360`. Thank for checking!

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+            logger.info(f'    left side (-180 -> +180): {check_1_str}')
+
+            # Right side of the antimeridian crossing: +180 -> +360
+            file_polygon_2 = _get_ogr_polygon(
```
same as above.

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+
+            # Right side of the antimeridian crossing: +180 -> +360
+            file_polygon_2 = _get_ogr_polygon(
+                max_x + ANTIMERIDIAN_CROSSING_RIGHT_SIDE_TEST_BUFFER,
```
Not sure where the constant `ANTIMERIDIAN_CROSSING_RIGHT_SIDE_TEST_BUFFER ` is defined. Can you check?

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +153,304 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+                logger.info(f'    {coverage_logger_str}:'
+                            ' Full (with antimeridian crossing')
+
+                # update DSWx-HLS product metadata
```
```suggestion
                # update RTC product metadata
```

seongsujeong reviewed (commented):
Here is my comments about the changes in `core.py` - Seongsu

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -268,13 +270,60 @@ def get_point_epsg(lat, lon):
+        err_str += f' be positive (x_spacing: {x_spacing})'
+        logger.error(err_str)
+        raise ValueError(err_str)
+    elif x_spacing is None and not epsg:
+        x_spacing = None
+    elif x_spacing is None and epsg == 4326:
+        x_spacing = 0.00027
+    elif x_spacing is None:
+        x_spacing = 30
```
```suggestion

    if x_spacing is None and epsg:
        x_spacing = 30
    elif x_spacing is None and epsg == 4326:
        x_spacing = 0.00027
```
Early return for the previous if statement in line 296 . Also no need to assign `None` when `x_spacing` is already `None`.

gshiroma replied:
Thank you, @seongsujeong . I think the logic above will fail if `epsg` is `4326` (we'd be assigning `30` to `x_spacing`), but I got the idea. Thank you!

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -268,13 +270,60 @@ def get_point_epsg(lat, lon):
+        err_str += f' be positive (y_spacing: {y_spacing_positive})'
+        logger.error(err_str)
+        raise ValueError(err_str)
+    elif y_spacing_positive is None and not epsg:
+        y_spacing = None
+    elif y_spacing_positive is None and epsg == 4326:
+        y_spacing = -0.00027
+    elif y_spacing_positive is None:
+        y_spacing = -30
+    else:
```
```suggestion

    if y_spacing_positive is None and epsg:
        y_spacing = -30
    elif y_spacing_positive is None and epsg == 4326:
        y_spacing = -0.00027
    else:
```

gshiroma replied:
Same here. Thanks

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -298,49 +345,67 @@ def generate_geogrids_from_db(bursts, geo_dict, dem, burst_db_file):
+    ymin_mosaic = mosaic_geogrid_dict['bottom_right']['y']
+    x_snap_mosaic = mosaic_geogrid_dict['x_snap']
+    y_snap_mosaic = mosaic_geogrid_dict['y_snap']
+
+
```
```suggestion
```
Nit pick. Leave the blank like as it is when you think it's absolutely necessary.

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -298,49 +345,67 @@ def generate_geogrids_from_db(bursts, geo_dict, dem, burst_db_file):
+    srs_bursts_majority = osr.SpatialReference()
+    srs_bursts_majority.ImportFromEPSG(epsg_bursts_majority)
+
+    if epsg_mosaic is None or epsg_mosaic == epsg_bursts_majority:
+        epsg_mosaic = epsg_bursts_majority
```
```suggestion
    if epsg_mosaic is None:
        epsg_mosaic = epsg_bursts_majority
```

gshiroma commented:
> Here is my comments about the changes in `core.py` - Seongsu

Thank you, @seongsujeong ! Just addressed them.

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -268,13 +270,56 @@ def get_point_epsg(lat, lon):
+        logger.error(err_str)
+        raise ValueError(err_str)
+    elif x_spacing is None and epsg and epsg == 4326:
+        x_spacing = 0.00027
```
Looks like we are repetitively using this constant, which I think is an approximate of `1.0 / 3600.0`. What do you think about defining this somewhere and using it? for example:

```python
SECONDS_IN_DEG = 1.0 / 3600.0
.
.
.
.
x_spacing = SECONDS_IN_DEG
```

gshiroma replied:
Thank you for this suggestion, @seongsujeong . I'm setting a constant, but I'm not using `1.0 / 3600.0`, because that number has a repeating decimal and I think it's better to avoid that for pixel spacing. I'll use something like `0.0002777`.

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -298,49 +341,65 @@ def generate_geogrids_from_db(bursts, geo_dict, dem, burst_db_file):
         Dict containing bursts geogrids indexed by burst_id
```
```suggestion
        Dict containing bursts' geogrids indexed by burst_id
```

seongsujeong reviewed (commented):
Here are my feedback about `geogrid.py` - Seongsu

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -52,18 +54,13 @@ def check_reprojection(geogrid_mosaic,
 
     for raster in rasters_to_check:
-        geotransform = raster.GetGeoTransform()
+        x0, dx, _, y0, _, dy = geotransform = raster.GetGeoTransform()
```
```suggestion
        x0, dx, _, y0, _, dy = raster.GetGeoTransform()
```
`geotransform` does not seem to be used in this function.

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
 
     if verbose:
-        print(f'mosaic dimension: {dim_mosaic}, #bands: {num_bands}')
+        print(f'    mosaic dimension: {dim_mosaic}, number of bands: {num_bands}')
```
What do you think about printing out the projection of the mosaic when in verbose mode?

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
+        else:
+
+            # calculate the burst RTC's offset wrt. the output mosaic in the image coordinate
+            offset_imgx = int((list_geo_transform[i,0] - xmin_mosaic) / posting_x + 0.5)
+            offset_imgy = int((list_geo_transform[i,3] - ymax_mosaic) / posting_y + 0.5)
```
```suggestion
            offset_imgx = int((list_geo_transform[i, 0] - xmin_mosaic) / posting_x + 0.5)
            offset_imgy = int((list_geo_transform[i, 3] - ymax_mosaic) / posting_y + 0.5)
```

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
-        #reshape arr_rtc when it is a singleband raster: to make it compatible in the for loop below
-        if num_bands==1:
-            arr_rtc=arr_rtc.reshape((1, arr_rtc.shape[0], arr_rtc.shape[1]))
+            band_ds = rtc_image_gdal_ds.GetRasterBand(i_band+1)
```
```suggestion
            band_ds = rtc_image_gdal_ds.GetRasterBand(i_band + 1)
```

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
+                                     offset_imgx: offset_imgx + width].copy()
+
+            if i_band == 0 and mosaic_mode.lower() == 'first':
+                ind = np.where(np.isnan(arr_temp))
```
```suggestion
                ind = np.isnan(arr_temp)
```
I think you can use [Boolean array indexing](https://numpy.org/doc/stable/user/basics.indexing.html#boolean-array-indexing) here

gshiroma replied:
These indexes are assigned in the first iteration and kept during the other iterations of the for loop. Using `np.where` would probably require less memory here.

seongsujeong replied:
in case of boolean array: `n * m` bytes where `n` and `m` are the # of rows and columns of the array respectively:
in case of np.where: `2 * 8 * p` bytes where `p` is the locations to the pixel

Therefore if the area with `nan` value is greater than 1/16 of the total area, then boolean indexing would be more memory efficient. Otherwise, `np.where` would be better.

This is the information I can provide you to decide which one to use. It's your call to decide which indexing you use.

gshiroma replied:
It makes sense, @seongsujeong . Yes, we'll have more than 1/16 of the image with NaNs, so just accepted your suggestion!

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
+
+                arr_distance_temp = arr_distance[offset_imgy: offset_imgy + length,
+                                                 offset_imgx: offset_imgx + width]
+                ind = np.where(np.logical_or(np.isnan(arr_distance_temp),
+                                             arr_new_distance <= arr_distance_temp))
```
```suggestion
                ind = np.logical_or(np.isnan(arr_distance_temp),
                                    arr_new_distance <= arr_distance_temp)
```

Same here.

gshiroma replied:
Please, see comment above.

seongsujeong replied:
https://github.com/opera-adt/RTC/pull/35#discussion_r1201197626

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
-        arr_numerator[i_band][valid_ind] = \
-            arr_numerator[i_band][valid_ind] / arr_denominator[valid_ind]
+    if mosaic_mode.lower() == 'average':
+        # Mode: average
```
```suggestion
        # Divide the numerator be denominator to calculate the average of the pixel intensity
```
For dummy proof like myself ;)

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
+    if mosaic_mode.lower() == 'average':
+        # Mode: average
+        for i_band in range(num_bands):
+            valid_ind = np.where(arr_denominator > 0)
```
```suggestion
            valid_ind = arr_denominator > 0
```
I think you can use boolean array indexing as the previous comments.

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
 
-        invalid_ind = np.where(arr_denominator == 0)
-        arr_numerator[i_band][invalid_ind] = np.nan
+            invalid_ind = np.where(arr_denominator == 0)
```
```suggestion
            invalid_ind = arr_denominator == 0
```
Same here

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -325,9 +532,9 @@ def compute_weighted_mosaic_raster(list_rtc_images, list_nlooks, geo_filename,
 
```
```suggestion
```

Another nit pick

gshiroma replied:
Thank you, @seongsujeong . I'm trying to use about 2 lines between functions.

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -325,9 +532,9 @@ def compute_weighted_mosaic_raster(list_rtc_images, list_nlooks, geo_filename,
+        scratch_dir='', geogrid_in=None, temp_files_list=None, verbose=True):
     '''
     Mosaic the snapped S1 geobursts
     paremeters:
```
```suggestion
    parameters:
```
- Also please remove `-----------` in the following line.
- `list_rtc` in the following docstring has to be `list_rtc_images`

seongsujeong replied:
Sorry that I confused to myself. It's a nit-pick but please remove `:` at the end of `parameters`.

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -338,14 +545,25 @@ def compute_weighted_mosaic_raster_single_band(
+            appended to this list
         verbose : bool
             Flag to enable/disable the verbose mode
+            
```
```suggestion

```
Removing white space here

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -271,26 +469,35 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
     Mosaic the snapped S1 geobursts
-    paremeters:
+
+    Parameters:
     -----------
         list_rtc: list
```
```suggestion
    Parameters:
        list_rtc_images: list
```

gshiroma replied:
We are trying to follow numpy style for docstrings: https://numpydoc.readthedocs.io/en/latest/format.html#parameters

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -271,26 +469,35 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
 
-def compute_weighted_mosaic_raster(list_rtc_images, list_nlooks, geo_filename,
-                                   geogrid_in=None, verbose=True):
+def mosaic_single_output_file(list_rtc_images, list_nlooks, geo_filename,
```
What do you think about changing `geo_filename` to `mosaic_filename` for better clarity?

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -325,9 +532,9 @@ def compute_weighted_mosaic_raster(list_rtc_images, list_nlooks, geo_filename,
+        list_rtc_images, list_nlooks, output_file_list, mosaic_mode,
+        scratch_dir='', geogrid_in=None, temp_files_list=None, verbose=True):
     '''
     Mosaic the snapped S1 geobursts
```
Please explicitly mention about the difference between this function and `mosaic_single_output_file`

seongsujeong reviewed (commented):
The comments for `mosaic_geobursts.py`. Most of them are minor things. - Seongsu

gshiroma commented on src/rtc/geogrid.py:
```diff
@@ -352,88 +411,115 @@ def generate_geogrids_from_db(bursts, geo_dict, dem, burst_db_file):
 
-        width = int(np.ceil((xmax - xmin) / x_spacing))
-        length = int(np.ceil((ymin - ymax) / y_spacing))
+        width = int(np.ceil((xmax - xmin) / x_spacing_bursts))
+        length = int(np.ceil((ymin - ymax) / y_spacing_bursts))
```
Sorry, @seongsujeong . But why do we need to add one? Please, notice we are already using `np.ceil()`.

seongsujeong replied:
I see. Thanks for the clarification.

gshiroma commented:
Thank you for your review, @seongsujeong . I think I addressed all your comments/concerns so far. Please let me know if I missed anything or if you'd have any further comments.

seongsujeong commented on src/rtc/h5_prep.py:
```diff
@@ -18,9 +18,7 @@
-BASE_HDF5_DATASET = f'/science/SENTINEL1'
-FREQ_GRID_SUB_PATH = 'RTC/grids/frequencyA'
-FREQ_GRID_DS = f'{BASE_HDF5_DATASET}/{FREQ_GRID_SUB_PATH}'
+FREQ_GRID_DS = '/data/frequencyA'
```
What do you think about adding a brief comment for what `FREQ_GRID_DS` stands for?

gshiroma replied:
@seongsujeong , for some reason, github is messing with these comments. Please, keep in mind that xmin and xmax are referenced to the edges of the pixel, not center. So, if you have 2 pixels and the pixel spacing is 30, the xmax - xmin will be 60

seongsujeong commented on src/rtc/h5_prep.py:
```diff
@@ -509,7 +513,7 @@ def get_metadata_dict(product_id: str,
+    metadata_dict['metadata/processingInformation/algorithms/radiometricTerrainCorrectionAlgorithmReference'] =\
         ['rtc_algorithm_reference',
          url_rtc_algorithm_document,
          'A reference to the RTC algorithm applied']
```
This is a suggestion. What do you think about adding information about what mosaic mode was used when `is_mosaic` is `True`?

gshiroma replied:
Thank you, @seongsujeong . This should be handled in a follow up PR . I'm separating the specs for the burst and mosaic products.

seongsujeong reviewed (commented):
Thanks for addressing my comments. Here are few more feedback from my end about the PR - Seongsu

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +156,327 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+    # within the [-180, +180] range, rather than the [0, +360] interval.
+    # This is verified by the test `min_x < -170`. There's no specific reason
+    # why -170 is used. It could be -160, or even 0.
+    flag_input_geographic_and_longitude_not_0_360 = \
+        file_srs.IsGeographic() and file_min_x < -170
```
```suggestion
    flag_input_geographic_and_longitude_not_0_360 = \
        file_srs.IsGeographic() and file_min_x < 0.0
```
If I understand this line correctly, `flag_input_geographic_and_longitude_not_0_360 ` will be`True` when `file_min_x` is out of the range [0, 360].
Let's assume some edge case that `file_min_x = -1.0`. then `flag_input_geographic_and_longitude_not_0_360` will be `False` because `file_min_x < -170` is `False`.
Please check if that is what you intended.

gshiroma replied:
Thank you, @seongsujeong . I prefer to use a number closer to `-180`, but not too close. The idea is that if the DEM domain is between, for example, [-160, 200] that would not cause a problem. The most common case is [0, 360], that's why I was using that range in the comments and variable name. But I updated the code and comments to make it clearer:

```
    # Flag to test if the ancillary input file is in geographic
    # coordinates and if its longitude domain is represented
    # within the [-180, +180] range, rather than, for example, inside
    # the [0, +360] interval.
    # This is verified by the test `min_x < -165`. There's no specific reason
    # why -165 is used. It could be -160, or even 0. However, testing for
    # -165 is more general than -160 or 0, but still not too close to -180.
    flag_input_geographic_and_longitude_lt_m165 = \
        file_srs.IsGeographic() and file_min_x < -165
```

seongsujeong commented on src/rtc/core.py:
```diff
@@ -152,6 +156,327 @@ def save_as_cog(filename, scratch_dir = '.', logger = None,
+
+        del ancillary_gdal_ds
+
+        ancillary_xf = ancillary_x0 + ancillary_width * ancillary_dx
+        ancillary_yf = ancillary_y0 + ancillary_length * ancillary_dy
```
Not so clear to me about what `f` means in `ancillary_xf` and `ancillary_yf`. Same for `geogrid_xf` and `geogrid_yf`.

gshiroma replied:
They are the final coordinates `[x0:xf]`; `[y0:yf]`. I added some comments to make this clearer

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -268,15 +273,58 @@ def get_point_epsg(lat, lon):
+    elif y_spacing_positive is None and epsg:
+        y_spacing = -30
+    elif y_spacing_positive is not None:
+        y_spacing = - y_spacing_positive
```
```suggestion
        y_spacing = -y_spacing_positive
```
Nit-pick here. PEP8 does not seem to mention about the spacing between the negation in unary operation, but I think it would be good to keep using the one style.
More information about what other language says about it:
https://softwareengineering.stackexchange.com/questions/189053/space-between-negative-sign-and-variable-name

seongsujeong commented on src/rtc/rtc_s1_single_job.py:
```diff
@@ -1190,8 +1209,8 @@ def run_single_job(cfg: RunConfig):
                 geo_burst_pol_filename = (f'NETCDF:{output_hdf5_file_burst}:'
-                                          '/science/SENTINEL1/RTC/grids/'
-                                          f'frequencyA/{pol}')
+                                          '/data/'
```
What do you think about importing `DATA_BASE_GROUP` from `h5_prep`, and use it instead of `/data`?

seongsujeong reviewed (commented):
Yet another comments. I think we are getting there. Nice work @gshiroma!

seongsujeong reviewed (approved):
I think this PR LGTM to merge. There is one last comment [here](https://github.com/opera-adt/RTC/pull/35#discussion_r1195702359) about how to index the invalid pixels, but I'll leave it to you to decide. Thanks for your hard work @gshiroma!

gshiroma commented:
> I think this PR LGTM to merge. There is one last comment [here](https://github.com/opera-adt/RTC/pull/35#discussion_r1195702359) about how to index the invalid pixels, but I'll leave it to you to decide. Thanks for your hard work @gshiroma!

Thank you, @seongsujeong . Just addressed that comment in this [commit](https://github.com/opera-adt/RTC/pull/35/commits/f2d10a45b04d5459f0d5cd0c901cb32cd13b595e). Should I wait for you to merge your PR in this one?
